My Jetpack: Add a Protect 'needs-attention' status when Threats are detected.

elliottprogrammer opened the pull request:
This PR adds a 'needs-attention' status to the Protect product card in My Jetpack when Protect threats are detected.
The Primary CTA changes to "Fix threats".
The 'needs-attention' status itself is actually hidden in the sense that the status still remains as "Active", however the product card border will either be yellow or red

With Threats | With Critical Threats
--- | ---
![Screen Shot 2024-12-16 at 09 05 01](https://github.com/user-attachments/assets/4a61f4dd-3630-4c52-83f2-0a97940c38ec) | ![Screen Shot on 2024-12-16 at 09-07-07](https://github.com/user-attachments/assets/8e559882-e43c-4947-94bc-639a1e5e1e78)


<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/jetpack-roadmap/issues/1939

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* I split the `needs_attention` status into 2 separate statuses (`needs-attention__warning` & `needs-attention__error`)
* When Protect detects threats, the protect card has yellow border (`needs-attention__warning`), if there are Critical threats, Protect has a red border (`needs-attention__error`).
* If threats are detected, the Protect CTA is "Fix threats".

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-bJN-p2
Design post: p1HpG7-umC-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- Install/activate Jetpack Protect with a Scan upgrade
- Using the Scan Helper in the Jetpack Debug Helper/Tools plugin, select the first EIACAR Threats selection to Add five EICAR threats to the site (one for every severity).
- Run a scan to detect them
- Proceed to the My Jetpack section, scroll down to the Protect product card.
- Verify the Protect product card has a red border, the CTA is "Fix threats" and the status appears as "Active".
- Now, using the Scan Helper in the Jetpack Debug Helper/Tools plugin again, remove the first EIACAR Threats selection  and add a different threat (like maybe a vulnerable plugin threat).
- Now verify the Protect product card has a yellow border, the CTA is "Fix threats" and the status appears as "Active".

elliottprogrammer committed:
Add red bubble alert if Backups failing.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Add 'needs-attention' status when Backup fails.

elliottprogrammer committed:
Remove Red bubble alert code, for use in followup PR.

elliottprogrammer committed:
Clean up, remove testing/dev code.

elliottprogrammer committed:
Fix doctype return value.

elliottprogrammer committed:
Phan suppress PhanTypeSuspiciousNonTraversableForeach warning.

elliottprogrammer committed:
Update changelog entry - incorrect description.

elliottprogrammer committed:
Add redbubble & notice when backups are failing.

elliottprogrammer committed:
Call Products\Backup class directly. fix phan TypeError.

elliottprogrammer committed:
Add changelog.

elliottprogrammer committed:
Adjust the red bubble notice priorities (reverse order).

elliottprogrammer committed:
Fix Notice styling for mobile view.

elliottprogrammer committed:
Fix last Backup date: Apply local timezone.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Add backup failed message in product card.

elliottprogrammer committed:
Merge branch 'trunk' into add/backup-needs-attention-status

elliottprogrammer committed:
Merge branch 'add/backup-needs-attention-status' into add/backup-redbubble-needs-attention-notice

elliottprogrammer committed:
Moar fixes to Notice mobile view.

elliottprogrammer committed:
Improve detection of failed backup.

elliottprogrammer committed:
Add 'source' property to needs-attention status data.

elliottprogrammer committed:
Protect product card: Fix Tooltips not displaying correctly in Threat Status.

elliottprogrammer committed:
Add a error/warning status and a "Fix threats" CTA when Protect has threats.

elliottprogrammer committed:
Merge branch 'trunk' into add/protect-fix-threats-status

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Remove scan/threats tooltip changes. (Changes completed in a seperate PR).

elliottprogrammer committed:
Merge branch 'trunk' into add/protect-fix-threats-status

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-card/action-button.tsx:
```diff
@@ -216,18 +216,62 @@ const ActionButton: FC< ActionButtonProps > = ( {
+					default:
+						return {
+							...defaultButton,
+						};
```
No need to de-struct the object here?

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-card/action-button.tsx:
```diff
@@ -216,18 +216,62 @@ const ActionButton: FC< ActionButtonProps > = ( {
+					label: __( 'Troubleshoot', 'jetpack-my-jetpack' ),
+					...( primaryActionOverride &&
+						PRODUCT_STATUSES.NEEDS_ATTENTION__ERROR in primaryActionOverride &&
+						primaryActionOverride[ PRODUCT_STATUSES.NEEDS_ATTENTION__ERROR ] ),
```
We could probably write

`...( primaryActionOverride?.[ PRODUCT_STATUSES.NEEDS_ATTENTION__ERROR ] ?? {} )`

We also avoid implicit conversion of falsy value in that way.

Not a blocker, more a fun fact.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-card/index.tsx:
```diff
@@ -219,6 +221,7 @@ const ProductCard: FC< ProductCardProps > = props => {
 						) }
 					</div>
 					<Status
+						slug={ slug }
```
I would avoid just passing down the slug, but rather add more context here as for what is it being used.

One idea would be to name the prop `suppressNeedsAttention={ slug === 'protect' }`.

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-card/status.tsx:
```diff
@@ -40,14 +45,18 @@ const getStatusLabel: StatusStateFunction = ( status, isOwned ) => {
+			if ( slug === 'protect' ) {
+				return __( 'Active', 'jetpack-my-jetpack' );
+			}
 			return __( 'Needs attention', 'jetpack-my-jetpack' );
```
I ask to understand more: are there no cases we need to actually display "Needs attention" in Protect card?

robertsreberski replied:
If so, I could make use of some comment here, cause it's a bit difficult to understand from the context.

elliottprogrammer replied:
> I ask to understand more: are there no cases we need to actually display "Needs attention" in Protect card?

As I understand it, for the Protect card, even when it has a NEEDS_ATTENTION__{WARNING | ERROR} status (which means threats have been detected), we still want to show the card status as 'Active' (per the designs: v1IWiVmOAvI8yT8CCa4b1I-fi-4398_9423).
I believe this is to portray to the user that Protect is still operating correctly/normally as it should (however threats were detected).
I left a comment in the code (as you suggested) to explain there also.

robertsreberski committed:
Fix i18n minification issue

robertsreberski reviewed (dismissed):
I have added one commit to fix minification issue with i18n. The code tests well after that!

I left some remarks, and the phan error should be addressed. Hope that will help move things further.

elliottprogrammer committed:
Maybe fix PhanTypeArraySuspicious TypeError.

elliottprogrammer committed:
Maybe fix PhanTypeArraySuspicious TypeError.

elliottprogrammer committed:
Add comment explaining Protect status exception, per feedback.

elliottprogrammer committed:
Merge branch 'trunk' into add/protect-fix-threats-status

elliottprogrammer committed:
Rename slug prop to productSlug, for more context.

elliottprogrammer committed:
Status card, rename prop to suppressNeedsAttention, per feedback.

elliottprogrammer commented:
> I left some remarks, and the phan error should be addressed. Hope that will help move things further.

@robertsreberski, thanks very much for the review! Much appreciated! 🙌
I've addressed all feedback & suggestions. Also the Phan error.
I believe everything should be good now. Let me know what you think! 🙂

elliottprogrammer committed:
Refactor the primaryActionOverride property code, per feedback.

elliottprogrammer committed:
Merge branch 'trunk' into add/protect-fix-threats-status

elliottprogrammer committed:
Fix method how active standalone plugin is detected.

elliottprogrammer committed:
Refactor 'status' destructure from redBubbleAlerts data.

IanRamosC reviewed (approved):
The change looks good and works as expected! :shipit:

robertsreberski reviewed (approved):
Ship it! Looks good to me, thanks for addressing everything, and including the comment! 😌

elliottprogrammer committed:
Merge branch 'trunk' into add/protect-fix-threats-status
